### PR TITLE
Add support for SMORMS3 optimizer

### PR DESCRIPTION
**Pull request by benob**
This PR adds support for the SMORMS3 optimizer described here:
http://sifter.org/~simon/journal/20150420.html
This optimizer yields faster convergence on certain workloads. The cost is higher memory requirements (stores 3 parameters per weight). A CUDA kernel is provided as well.

### CODE REVIEW

File: chainer/optimizers/smorms3.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+    """ Simon Funk's SMORMS3, see http://sifter.org/~simon/journal/20150420.html """
+
+    def __init__(self, lr=0.001, eps=1e-16):
+        self.lr = lr
+        self.eps = eps
+
+    def init_state(self, param, state):
+        xp = cuda.get_array_module(param.data)
+        state['mem'] = xp.zeros_like(param.data) + 1
```
Comment (okuta) [inline chainer/optimizers/smorms3.py]: How about using `ones_like`?

Comment (benob) [inline chainer/optimizers/smorms3.py]: Working on a version that matches chainer/master.

**Comment by benob** [comment]
Squashed commits for cleaner history (maybe that was not a good idea).
Tagging with WIP because I could not test the GPU code.

**Comment by benob** [comment]
I don't understand why the checks failed.

**Comment by delta2323** [comment]
I do not know why CI does not work. Adding new commit could trigger the CI. As this branch has conflicts, how about adding the commit that merges or rebases master branch?

**Comment by delta2323** [comment]
Thank you for your PR. Could you tell me about this optimizer as I do not know about it well? Specifically, is there a published paper for it and how much does it used?

**Comment by benob** [comment]
SMORMS3 is described here: http://sifter.org/~simon/journal/20150420.html
I haven't seen it described in a proper paper. However, some people use it around the web and it is mentioned in a few papers. There is a theano implementation:
https://github.com/diogo149/treeano/blob/master/treeano/sandbox/nodes/smorms3.py

Feel free to include it or not in chainer. I figured it could be interesting to its users.

**Comment by delta2323** [comment]
>I haven't seen it described in a proper paper. However, some people use it around the web and it is mentioned in a few papers. There is a theano implementation:

OK, I understand it. In my opinion, it is acceptable to add the optimizer.

**Comment by delta2323** [comment]
Please register this optimizer to `docs/source/reference/optimizers.rst`.

File: chainer/optimizers/smorms3.py (hunk 1 above)
Comment (delta2323) [inline chainer/optimizers/smorms3.py]: Could you apply this change?

**Comment by benob** [comment]
Registered the optimizer and fixed the ones_like(...) line.

**Comment by delta2323** [comment]
LGTM 👍

**Comment by delta2323** [comment]
Thank you very much!

**Pull request merged**